astropy/timeseries/tests/test_downsample.py

```diff
@@ -35,35 +37,104 @@ def test_timeseries_invalid():
     assert_equal(down_3.time_bin_start.isot, Time(['2016-03-22T12:30:31.000', '2016-03-22T12:30:34.000']))
-    assert_equal(down_3["a"].data, np.array([2, 4]))
+    assert_equal(down_3["a"].data.data, np.array([2, 4]))
 
     down_4 = aggregate_downsample(ts, time_bin_size=4*u.second)
-    u.isclose(down_4.time_bin_size, [4]*u.second)
-    assert_equal(down_4.time_bin_start.isot, Time(['2016-03-22T12:30:31.000']))
-    assert_equal(down_4["a"].data, np.array([2]))
+    u.isclose(down_4.time_bin_size, [4, 4]*u.second)
+    assert_equal(down_4.time_bin_start.isot, Time(['2016-03-22T12:30:31.000', '2016-03-22T12:30:35.000']))
+    assert_equal(down_4["a"].data.data, np.array([2, 5]))
 
     down_units = aggregate_downsample(ts_units, time_bin_size=4*u.second)
-    u.isclose(down_units.time_bin_size, [4]*u.second)
-    assert_equal(down_units.time_bin_start.isot, Time(['2016-03-22T12:30:31.000']))
+    u.isclose(down_units.time_bin_size, [4, 4]*u.second)
+    assert_equal(down_units.time_bin_start.isot, Time(['2016-03-22T12:30:31.000', '2016-03-22T12:30:35.000']))
     assert down_units["a"].unit.name == 'ct'
-    assert_equal(down_units["a"].data, np.array([2.5]))
+    assert_equal(down_units["a"].data, np.array([2.5, 5.0]))
+
+    # Uneven bin sizes
+    down_uneven_bins = aggregate_downsample(ts, time_bin_size=[2, 1, 1]*u.second)
+    u.isclose(down_uneven_bins.time_bin_size, [2, 1, 1]*u.second)
+    assert_equal(down_uneven_bins.time_bin_start.isot, Time(['2016-03-22T12:30:31.000',
+                                                             '2016-03-22T12:30:33.000',
+                                                             '2016-03-22T12:30:34.000']))
+    assert_equal(down_uneven_bins["a"].data.data, np.array([1, 3, 4]))
+
+    # Uneven bins
+    down_time_array = aggregate_downsample(ts, time_bin_start=Time(['2016-03-22T12:30:31.000',
+                                                                    '2016-03-22T12:30:33.000']),
+                                           time_bin_end=Time(['2016-03-22T12:30:32.000',
+                                                              '2016-03-22T12:30:36.000']))
+    u.isclose(down_time_array.time_bin_size, [1, 2]*u.second)
+    assert_equal(down_time_array.time_bin_start.isot, Time(['2016-03-22T12:30:31.000',
+                                                            '2016-03-22T12:30:33.000']))
+    assert_equal(down_time_array["a"].data.data, np.array([1, 4]))
+
+    # Overlapping bins
+    with pytest.warns(AstropyUserWarning, match="Overlapping bins should be avoided since they "
+                                                "can lead to double-counting of data during binning."):
+        down_overlap_bins = aggregate_downsample(ts, time_bin_start=Time(['2016-03-22T12:30:31.000',
+                                                                          '2016-03-22T12:30:33.000']),
+                                                 time_bin_end=Time(['2016-03-22T12:30:34.000',
```

Comment:
```suggestion
                                                 time_bin_end=Time(['2016-03-22T12:30:33.999,
```
needs to stay `< '2016-03-22T12:30:34.000'` since `ts.time[3]` changed above.